tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -3811,14 +3811,24 @@ SpirvEmitter::processByteAddressBufferStructuredBufferGetDimensions(
 SpirvInstruction *SpirvEmitter::processRWByteAddressBufferAtomicMethods(
     hlsl::IntrinsicOp opcode, const CXXMemberCallExpr *expr) {
   // The signature of RWByteAddressBuffer atomic methods are largely:
-  // void Interlocked*(in UINT dest, in UINT value);
-  // void Interlocked*(in UINT dest, in UINT value, out UINT original_value);
+  // void Interlocked*(in UINT dest, in T value);
+  // void Interlocked*(in UINT dest, in T value, out T original_value);
+
   const auto *object = expr->getImplicitObjectArgument();
   auto *objectInfo = loadIfAliasVarRef(object);
 
   auto *zero =
       spvBuilder.getConstantInt(astContext.UnsignedIntTy, llvm::APInt(32, 0));
-  auto *offset = doExpr(expr->getArg(0));
+  const auto *dest = expr->getArg(0);
+  const auto baseType = dest->getType()->getCanonicalTypeUnqualified();
```

Comment:
Looking at the DXIL generated for these cases, the type of the operation is determined by the type of `value`, not the type of `dest`, which is actually an offset.